### PR TITLE
Feature/department as model

### PR DESCRIPTION
**Pull request by gecBurton**
image

### CODE REVIEW

File: one_big_thing/learning/constants.py [hunk 1]
```diff
@@ -37,36 +37,6 @@
 BL_LINK = "https://wiki.bl.uk:8443/display/DSSP/Calendar+of+Events"
 
 
-# These are departments having the "streamlined view"
-# Includes arms' length bodies that use main dept's intranet
-# These keys are in a particular format to link to other lists
-DEPARTMENTS_USING_INTRANET_LINKS = {
```
Comment (gecBurton) [inline one_big_thing/learning/constants.py:43]: i have moved this into a model and database migration

File: one_big_thing/learning/models.py [hunk 2]
```diff
@@ -30,6 +30,81 @@ class Meta:
...
+        (WELSH_GOVERNMENT, "Welsh Government"),
+    ]
+
+    code = models.CharField(max_length=128, unique=True, help_text="unique code for department")
+    display = models.CharField(max_length=128, help_text="display name")
+    parent = models.CharField(max_length=128, choices=PARENTS, help_text="parent department")
+    url = models.URLField(null=True, blank=True, help_text="intranet link")
+
+    @classmethod
+    def choices(cls) -> list[tuple[Any, Any]]:
```
Comment (gecBurton) [inline one_big_thing/learning/models.py:104]: we could cache this?

File: one_big_thing/learning/schemas.py [hunk 3]
```diff
@@ -181,3 +181,10 @@ class Meta:
             "required": "You must select a profession",
         },
     )
+
+    def validate_department(self, value):
```
Comment (gecBurton) [inline one_big_thing/learning/schemas.py:185]: now that the departments aren't static we need to make this validation dynamic

File: tests/test_totp_admin.py [hunk 4]
```diff
@@ -44,6 +44,6 @@ def test_admin_login_fail(self):
         )
 
         # Check if the login was unsuccessful (bizarrely this is an HTTP 302 status code)
-        self.assertEqual(response.status_code, 200)
+        self.assertEqual(response.status_code, 302)
```
Comment (gecBurton) [inline tests/test_totp_admin.py:47]: oops! this change inadvertently corrects an error

File: one_big_thing/learning/models.py [hunk 5]
```diff
@@ -30,6 +30,84 @@ class Meta:
         abstract = True
 
 
+class Department(TimeStampedModel):
+    ATTORNEY_GENERALS_DEPARTMENTS = "ATTORNEY_GENERALS_DEPARTMENTS"
```
Comment (gecBurton) [inline one_big_thing/learning/models.py:34]: this list is from 10DS

File: one_big_thing/learning/admin.py [hunk 6]
```diff
@@ -4,30 +4,40 @@
...
 class UserAdmin(admin.ModelAdmin):
     search_fields = ("email", "first_name", "last_name")
+    list_display = ("email", "department")
 
 
 class OTPAdmin(OTPAdminSite):
     pass
 
 
-admin_site = OTPAdmin(name="OTPAdmin")
```
Comment (gecBurton) [inline one_big_thing/learning/admin.py:24]: we want to use the normal admin locally

**Comment by nmenezes0** [comment]
> image

Looks like the "Other" depts are incorrectly given the parent dept as "Defra".

File: one_big_thing/learning/models.py [hunk 7]
```diff
@@ -30,6 +30,84 @@ class Meta:
...
+    CABINET_OFFICE = "CABINET_OFFICE"
+    CHANCELLORS_OTHER_DEPARTMENTS = "CHANCELLORS_OTHER_DEPARTMENTS"
+    CHARITY_COMMISSION = "CHARITY_COMMISSION"
+    COMPETITION_AND_MARKETS_AUTHORITY = "COMPETITION_AND_MARKETS_AUTHORITY"
+    CROWN_PROSECUTION_SERVICE = "CROWN_PROSECUTION_SERVICE"
+    DEPARTMENT_FOR_BUSINESS_AND_TRADE_ = "DEPARTMENT_FOR_BUSINESS_AND_TRADE_"
```
Comment (nmenezes0) [inline one_big_thing/learning/models.py:40]: Is there a reason for the underscore after the name? (Not that it particularly makes a diff.)

File: departments.csv [hunk 8]
```diff
@@ -0,0 +1,177 @@
+"code", "display", "parent", "url"
+"acas", "The Advisory, Conciliation and Arbitration Service (Acas)", "DEPARTMENT_FOR_ENVIRONMENT_FOOD_AND_RURAL_AFFAIRS", null
```
Comment (nmenezes0) [inline departments.csv:2]: Is this right for Acas? https://www.gov.uk/government/organisations/acas
Also don't think it's right for "queen-elizabeth-ii-conference-centre", "Queen Elizabeth II Conference Centre (QEIICC)" (arms' length body of DLUCH)

Looks like something might have gone wrong when compiling the list and DEFRA keeps appearing as parent org.

**Comment by nmenezes0** [comment]
Haven't tested or reviewed in detail - but approach looks fine to me. Just a few possible data issues have snuck in there (see comments).

File: one_big_thing/learning/models.py (hunk 7 above)
Comment (gecBurton) [reply one_big_thing/learning/models.py:40]: no reason. this is a typo

File: departments.csv (hunk 8 above)
Comment (gecBurton) [reply departments.csv:2]: good point!

Comment (gecBurton) [reply departments.csv]: [78df662](https://github.com/i-dot-ai/one-big-thing/pull/270/commits/78df662a8c115f5e78b91cd50b2a114673c8739e)

File: pytest_tests/test_migrations.py [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+    "yorkshire-dales-national-park-authority",
+    "other",
+    "other-civil-servant",
+    "other-public-servant",
+]
+
+
+@pytest.mark.django_db
+def test_0020_department():
+    migrator = Migrator(database="default")
```
Comment (nmenezes0) [inline pytest_tests/test_migrations.py:186]: This is fun, I've never tested a migration before.

File: one_big_thing/learning/models.py [hunk 10]
```diff
@@ -45,7 +120,8 @@ class User(BaseUser, UUIDPrimaryKeyBase):
     invited_at = models.DateTimeField(default=None, blank=True, null=True)
     invite_accepted_at = models.DateTimeField(default=None, blank=True, null=True)
     last_token_sent_at = models.DateTimeField(editable=False, blank=True, null=True)
-    department = models.CharField(max_length=254, blank=True, null=True, choices=department_tuples)
+    old_department = models.CharField(max_length=254, blank=True, null=True)
```
[the remark this answers is not in the document]
Comment (nmenezes0) [reply one_big_thing/learning/models.py:148]: Maybe in a separate PR once the changes have been deployed and we're happy?

File: one_big_thing/learning/schemas.py (hunk 3 above)
Comment (nmenezes0) [reply one_big_thing/learning/schemas.py:185]: Need a decorator `@validates("department")` - https://marshmallow.readthedocs.io/en/stable/quickstart.html#field-validators-as-methods

Comment (nmenezes0) [reply one_big_thing/learning/schemas.py:185]: Test if you want to use it (I was writing one so I could test):
```
def test_my_details_schema():
    details_schema = schemas.MyDetailsSchema()
    invalid = {"department": "dodgy-dept", "grade": "GRADE7", "profession": "ANALYSIS"}
    valid = {"department": "cabinet-office", "grade": "GRADE7", "profession": "ANALYSIS"}
    validation_issues = details_schema.validate(invalid)
    assert "department" in validation_issues, validation_issues
    validation_issues = details_schema.validate(valid)
    assert not validation_issues, validation_issues
```

**Comment by nmenezes0** [comment]
In the "My details page" (http://localhost:8055/my-details/), the department is always displaying as Acas. Have checked in the DB and the correct department is save (the correct department having been saved previously, and that one wasn't Acas).

Not sure what's going on here - something in the view/template?

File: one_big_thing/learning/migrations/departments.csv [hunk 11]
```diff
@@ -0,0 +1,177 @@
+"code", "display", "parent", "url"
```
Comment (nmenezes0) [inline one_big_thing/learning/migrations/departments.csv:1]: Assume there's only a URL for streamlined departments? Is it worth changing the title of the column (here, or in the model) as there are also other departmental URLs flying around?

Not necessary change - just a thought.

**Review by nmenezes0 — CHANGES_REQUESTED** [review]
Only two things that need to be changed (there are comments in the PR):
1. The validation for departments
2. The issue where the my details form isn't being populated properly for the department

Otherwise looks good - other comments are minor, or questions for my understanding.

**Comment by gecBurton** [comment]
> In the "My details page" (http://localhost:8055/my-details/), the department is always displaying as Acas. Have checked in the DB and the correct department is save (the correct department having been saved previously, and that one wasn't Acas).
> 
> Not sure what's going on here - something in the view/template?
> 
> image

good spot! fixed here [658e9af](https://github.com/i-dot-ai/one-big-thing/pull/270/commits/658e9af8a1ca3aca2ca4b402ac74aa9144488add)

File: one_big_thing/learning/schemas.py (hunk 3 above)
Comment (nmenezes0) [reply one_big_thing/learning/schemas.py:187]: @gecBurton - has this validation been fixed too?

**Review by nmenezes0 — APPROVED** [review]

**Pull request merged**